File: 13.Linked_List/python/4.ReverseLinkedList.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

# Manually create the linked list
head = Node(1)
second = Node(2)
third = Node(3)
four = Node(4)
five = Node(5)

# Link the nodes
head.next = second
second.next = third
third.next = four
four.next = five

# Reverse function
def reverse(head):
    prev = None
    curr = head
    while curr is not None:
        next_node = curr.next
        curr.next = prev
        prev = curr
        curr = next_node
    return prev

# Reverse the list
head = reverse(head)

# Print the reversed list
current = head
result = ""
while current is not None:
    result += str(current.data) + " -> "
    current = current.next
result += "None"

print(result)

File: 13.Linked_List/javascript/4.ReverseLinkedList.js
class Node {
    constructor(data) {
        this.data = data
        this.next = null
    }
}

function buildLinkedList(arr) {
    if (arr.length === 0) return null

    let head = new Node(arr[0])
    let current = head

    for (let i = 1; i < arr.length; i++) {
        current.next = new Node(arr[i])
        current = current.next
    }

    return head
}

class Solution {
    reverseNodes(head) {
        let current = head
        let prev = null

        while (current !== null) {
            let next = current.next
            current.next = prev
            prev = current
            current = next
        }

        return prev
    }


}

function PrintList(head) {
    let res = ''
    let current = head
    while (current !== null) {
        res += current.data + ' '
        current = current.next
    }

    console.log(res)
}

let arr = [1, 2, 3, 4, 5]
let n = 3
let Sol = new Solution()

let head = buildLinkedList(arr)
let newHead = Sol.reverseNodes(head)
PrintList(newHead)
